### PR TITLE
Allow CLI to pass environment variables to FOD builder

### PR DESCRIPTION
**Pull request by Ninlives**
# Motivation

This pull request is based on #5890. As suggested, this option will now only set the environment of builders of fixed-output derivations, and only those variables listed in `impureEnvVars` will take effect.

# Context

Based on: #5890

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libstore/globals.cc
Comment (thufschmitt) [inline src/libstore/globals.cc:121]: This is a bit irky, and also somewhat wrong if the client config doesn't include the daemon config (for instance when building on a remote machine): Even if the client config doesn't specify the `environment` setting, its default value will be sent and override whatever the daemon might have set.

Do we really need the “get from the environment if not specified” logic when parsing the `environment` configuration option? I'd rather stay simple and predictable (at least for a first version) and just have this variable be a list of key-value pairs.

File: src/libstore/globals.hh
Comment (thufschmitt) [inline src/libstore/globals.hh:1050]: I'm not sure the name is right any more now that this setting doesn't blindly change the environment of the daemon.
Maybe something like `impure-env-var-overrides` to link it more explicitly to the `impureEnvVar` option? Probably that should be also reflected in the documentation, and there should be a reference from the documentation of `impureEnvVar` to this setting.

**Review by thufschmitt — CHANGES_REQUESTED** [review]
Thanks,

I think this is good overall, however I'm not keen on having the magic substitution by the environment variable. It is definitely handy, but it also makes the semantics rather complex, so I think we should leave it out for now.

**Comment by Ninlives** [comment]
Done. I don't quite like the option name, but I haven't found a better alternative :)

**Comment by thufschmitt** [comment]
Thanks @Ninlives!

@NixOS/nix-team, is the `impure-env-vars-override` name good for you? If I get a green light before next week I'll merge this straight away, otherwise I'll merge it as experimental until we have an agreement on the name.

**Comment by roberth** [comment]
I think `impure-env` would be sufficient. It makes sense to me when `impureEnvVars` picks from `impure-env` first and the actual environment variables as a fallback.

[impureEnvVars docs](https://nixos.org/manual/nix/stable/language/advanced-attributes.html?highlight=impureenv#adv-attr-impureEnvVars) need to be updated regardless.

**Comment by Ericson2314** [comment]
I think this is a great future, but I am sort of disgusted by *older feature*, the fact that arbitrary env vars can be snooped with `impureEnvVars`.

I think it would be nice to sandbox that, and I think it makes sense for that to be part of an `impure-env` setting:

1. Today's permissiveness
    ```
    impure-env = *=inherit
    ```

2. Set some variables, whitelist two others to be inherited
    ```
    impure-env = foo=x bar=y baz=inherit quux=inherit
    ```

3. Set some variables, allow the rest to be inherited
4.     ```
    impure-env = foo=x bar=y *=inherit
    ```

**Comment by fricklerhandwerk** [comment]
Triaged in the Nix team meeting 2023-09-29:

- @thufschmitt: the main difference to existing behavior is that
    1. you can now pass impure env vars through the CLI
    2. you don't have to be root to do that
- @edolstra: this option seems hard to use
    - you'd probably want to inherit variables from the call site
    - @thufschmitt: this is what was done originally, but it would create a lot of potential for confusion
        - it felt safer to start with the simple solution
        - the clever way would also make the config parsing rather weird
        - we could still add it as a follow-up
- @tomberek: would this give you variables you wouldn't have otherwise?
    - @thufschmitt: no, nothing you wouldn't be able to craft in the language
- @ericson2314: this should also restrict access to the specified variables
    - could add this as experimental to avoid front-loaded bikeshedding
- decision: assigned to @thufschmitt, merge as experimental feature to iterate

**Comment by thufschmitt** [comment]
Applied the suggestions from the @NixOS/nix-team. Merging as experimental

**Comment by SuperSandro2000** [comment]
I shouldn't be getting those log lines without actually configuring impure-env and having it on empty string, right?

```
warning: Ignoring setting 'impure-env' because experimental feature 'configurable-impure-env' is not enabled
```

```
 ▶ nix show-config | rg impure-env
impure-env =
```

**Pull request merged**
